README.md
# Node Res

> A facade over Node.js HTTP `res` object with no side-effects.

[![NPM Version][npm-image]][npm-url]
[![Build Status][travis-image]][travis-url]
[![Appveyor][appveyor-image]][appveyor-url]
[![Coveralls][coveralls-image]][coveralls-url]

<a href="http://i1117.photobucket.com/albums/k594/thetutlage/poppins-1_zpsg867sqyl.png">
<img src="http://i1117.photobucket.com/albums/k594/thetutlage/poppins-1_zpsg867sqyl.png" width="300px" align="right" vspace="20px" />
</a>

`node-res` is a simple module to make HTTP response in Node.js. In offers helpers to make it easier to set `headers`, define response statuses and properly parse response type to set appropriate headers.

For example:

```js
// content-type: plain/text
nodeRes.send(req, res, 'Hello world')

// content-type: application/json
nodeRes.send(req, res, { greeting: 'hello world' })
    
// content-type: text/html
nodeRes.send(req, res, '<h2> Hello world </h2>')
```

## See also

1. [node-req](https://npmjs.org/package/node-req)
2. [node-cookie](https://npmjs.org/package/node-cookie)

## Responding to requests.

```javascript
const http = require('http')
const nodeRes = require('node-res')

http.createServer(function (req, res) {
  
  // plain text
  nodeRes.send(req, res, "Hello world")

  // json
  nodeRes.json(req, res, {time:"now"})

  // jsonp
  nodeRes.jsonp(req, res, {time:"now"}, "callback")

}).listen(3000)

```

nodeRes takes http server `res` object as first argument to perform any operations.

## Methods

### getHeader
Returns the value of an existing header on
the response object

**Params**

| Param | Type | Required | Description |
|-----|-------|------|------|
| res | Object | Yes | &nbsp; |
| key | String | Yes | &nbsp; |

**Returns**
Array

**Example**
```js
nodeRes.getHeader(res, 'Content-type')
```

----
### header
Sets header on the response object

**Params**

| Param | Type | Required | Description |
|-----|-------|------|------|
| res | Object | Yes | &nbsp; |
| key | String | Yes | &nbsp; |
| value | String | Yes | &nbsp; |

**Returns**
Void

**Example**
```js
nodeRes.header(res, 'Content-type', 'application/json')

// or set an array of headers
nodeRes.header(res, 'Link', ['<http://localhost/>', '<http://localhost:3000/>'])
```

----
### status
Set status on the HTTP res object

**Params**

| Param | Type | Required | Description |
|-----|-------|------|------|
| res | Object | Yes | &nbsp; |
| code | Number | Yes | &nbsp; |

**Returns**
Void

**Example**
```js
nodeRes.status(200)
```

----
### safeHeader
Sets the header on response object, only if it
does not exists.

**Params**

| Param | Type | Required | Description |
|-----|-------|------|------|
| res | Object | Yes | &nbsp; |
| key | String | Yes | &nbsp; |
| value | String | Yes | &nbsp; |

**Returns**
Void

**Example**
```js
nodeRes.safeHeader(res, 'Content-type', 'application/json')
```

----
### removeHeader
removing header using it's key

**Params**

| Param | Type | Required | Description |
|-----|-------|------|------|
| res | Object | Yes | &nbsp; |
| key | String | Yes | &nbsp; |

**Returns**
Void

----
### write
Write string or buffer to the response object.

**Params**

| Param | Type | Required | Description |
|-----|-------|------|------|
| res | Object | Yes | &nbsp; |
| body | String|Buffer | Yes | &nbsp; |

**Returns**
Void

**Example**
```js
nodeRes.write(res, 'Hello world')
```

----
### end
Explictly end HTTP response

**Params**

| Param | Type | Required | Description |
|-----|-------|------|------|
| res | Object | Yes | &nbsp; |

**Returns**
Void

----
### send
Send body as the HTTP response and end it. Also
this method will set the appropriate `Content-type`
and `Content-length`.

If body is set to null, this method will end the response
as 204.

**Params**

| Param | Type | Required | Description |
|-----|-------|------|------|
| req | Object | Yes | &nbsp; |
| res | Object | Yes | &nbsp; |
| body | Mixed | Yes | &nbsp; |

**Returns**
Void

**Example**
```js
nodeRes.send(req, res, 'Hello world')

// or html
nodeRes.send(req, res, '<h2> Hello world </h2>')

// or JSON
nodeRes.send(req, res, { greeting: 'Hello world' })

// or Buffer
nodeRes.send(req, res, Buffer.from('Hello world', 'utf-8'))
```

----
### json
Returns the HTTP response with `Content-type`
set to `application/json`.

**Params**

| Param | Type | Required | Description |
|-----|-------|------|------|
| req | Object | Yes | &nbsp; |
| res | Object | Yes | &nbsp; |
| body | Object | Yes | &nbsp; |

**Returns**
Void

**Example**
```js
nodeRes.json(req, res, { name: 'virk' })
nodeRes.json(req, res, [ 'virk', 'joe' ])
```

----
### jsonp
Make JSONP response with `Content-type` set to
`text/javascript`.

**Params**

| Param | Type | Required | Description |
|-----|-------|------|------|
| req | Object | Yes | &nbsp; |
| res | Object | Yes | &nbsp; |
| body | Object | Yes | &nbsp; |
| callbackFn  | String | No | &nbsp; |

**Returns**
Void

**Example**
```js
nodeRes.jsonp(req, res, { name: 'virk' }, 'callback')
```

----
### download
Download file as a stream. Stream will be closed once
download is finished.

Options are passed directly to [send](https://www.npmjs.com/package/send)

**Params**

| Param | Type | Required | Description |
|-----|-------|------|------|
| req | Object | Yes | &nbsp; |
| res | Object | Yes | &nbsp; |
| filePath | String | Yes | &nbsp; |
| options  | Object | No | &nbsp; |

**Returns**
Void

**Example**
```js
nodeRes.download(req, res, '/storage/data.txt')
```

----
### attachment
Send file as a stream with Content-Disposition of attachment
which forces the download of the file.

**Params**

| Param | Type | Required | Description |
|-----|-------|------|------|
| req | Object | Yes | &nbsp; |
| res | Object | Yes | &nbsp; |
| filePath | String | Yes | &nbsp; |
| name  | String | No | &nbsp; |
| disposition  | String | No | &nbsp; |
| options | Object | No | &nbsp; |

**Returns**
Void

**Example**
```js
nodeRes.attachment(req, res, '/storage/data.txt', 'data.txt')
```

----
### location
Set `Location` header on the HTTP response.

**Params**

| Param | Type | Required | Description |
|-----|-------|------|------|
| res | Object | Yes | &nbsp; |
| url | String | Yes | &nbsp; |

**Returns**
Void

----
### redirect
Redirect the HTTP request to the given url.

**Params**

| Param | Type | Required | Description |
|-----|-------|------|------|
| req | Object | Yes | &nbsp; |
| res | Object | Yes | &nbsp; |
| url | String | Yes | &nbsp; |
| status  | Number | No | &nbsp; |

**Returns**
Void

**Example**
```js
nodeRes.redirect(req, res, '/')
```

----
### vary
Add vary header to the HTTP response.

**Params**

| Param | Type | Required | Description |
|-----|-------|------|------|
| res | Object | Yes | &nbsp; |
| field | String | Yes | &nbsp; |

**Returns**
Void

----
### type
Set content type header by looking up the actual
type and setting charset to utf8

**Params**

| Param | Type | Required | Description |
|-----|-------|------|------|
| res | Object | Yes | &nbsp; |
| type | String | Yes | &nbsp; |
| charset  | String | No | &nbsp; |

**Returns**
Void

**Example**
```js
nodeRes.type(res, 'html')
nodeRes.type(res, 'json')
nodeRes.type(res, 'application/json')
```

----


## Descriptive methods
Node res also has support for descriptive methods, they set the status itself without calling the `status` method.

```javascript
nodeRes.ok(req, res, 'Hello world') // will set 200 as status
nodeRes.unauthorized(req, res, 'You must login first') // will set 401 as status
```

| method | http response status |
|--------|-------------|
| continue | 100 |
| switchingProtocols | 101 |
| ok | 200 |
| created | 201 |
| accepted | 202 |
| nonAuthoritativeInformation | 203 |
| noContent | 204 |
| resetContent | 205 |
| partialContent | 206 |
| multipleChoices | 300 |
| movedPermanently | 301 |
| found | 302 |
| seeOther | 303 |
| notModified | 304 |
| useProxy | 305 |
| temporaryRedirect | 307 |
| badRequest | 400 |
| unauthorized | 401 |
| paymentRequired | 402 |
| forbidden | 403 |
| notFound | 404 |
| methodNotAllowed | 405 |
| notAcceptable | 406 |
| proxyAuthenticationRequired | 407 |
| requestTimeout | 408 |
| conflict | 409 |
| gone | 410 |
| lengthRequired | 411 |
| preconditionFailed | 412 |
| requestEntityTooLarge | 413 |
| requestUriTooLong | 414 |
| unsupportedMediaType | 415 |
| requestedRangeNotSatisfiable | 416 |
| expectationFailed | 417 |
| unprocessableEntity | 422 |
| tooManyRequests | 429 |
| internalServerError | 500 |
| notImplemented | 501 |
| badGateway | 502 |
| serviceUnavailable | 503 |
| gatewayTimeout | 504 |
| httpVersionNotSupported | 505 |


[appveyor-image]: https://img.shields.io/appveyor/ci/thetutlage/node-res/master.svg?style=flat-square

[appveyor-url]: https://ci.appveyor.com/project/thetutlage/node-res

[npm-image]: https://img.shields.io/npm/v/node-res.svg?style=flat-square
[npm-url]: https://npmjs.org/package/node-res

[travis-image]: https://img.shields.io/travis/poppinss/node-res/master.svg?style=flat-square
[travis-url]: https://travis-ci.org/poppinss/node-res

[coveralls-image]: https://img.shields.io/coveralls/poppinss/node-res/develop.svg?style=flat-square

[coveralls-url]: https://coveralls.io/github/poppinss/node-res


CHANGELOG.md
<a name="3.0.2"></a>
## [3.0.2](https://github.com/poppinss/node-res/compare/v3.0.1...v3.0.2) (2017-06-12)



<a name="3.0.1"></a>
## [3.0.1](https://github.com/poppinss/node-res/compare/v2.0.1...v3.0.1) (2016-11-08)


### Bug Fixes

* **response:** set charset with content type ([19d6b87](https://github.com/poppinss/node-res/commit/19d6b87)), closes [adonis-framework#329](https://github.com/adonis-framework/issues/329)


<a name="3.0.0"></a>
# [3.0.0](https://github.com/poppinss/node-res/compare/v2.0.1...v3.0.0) (2016-06-25)


### Bug Fixes

* **download:** use send module to download file([7134e51](https://github.com/poppinss/node-res/commit/7134e51))



<a name="2.0.2"></a>
## [2.0.2](https://github.com/poppinss/node-res/compare/v2.0.1...v2.0.2) (2016-01-19)




<a name="2.0.1"></a>
## 2.0.1 (2016-01-18)


### Features

* **package:** added commitizen for consistent commit messages ([bc0241c](https://github.com/poppinss/node-res/commit/bc0241c))





test/response.spec.js
'use strict'

/*
 * node-res
 *
 * (c) Harminder Virk <virk@adonisjs.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
*/

const test = require('japa')
const supertest = require('supertest')
const http = require('http')
const path = require('path')
const methods = require('../src/Response/methods')
const Response = require('../')

test.group('Response', function (assert) {
  test('should set dynamic method on response object which auto set statuses', async function (assert) {
    const server = http.createServer(function (req, res) {
      Response.tooManyRequests(req, res, 'Please wait')
    })
    const response = await supertest(server).get('/').expect(429)
    assert.equal(response.text, 'Please wait')
  })

  test('should have all descriptive methods on response status, point to right status', function (assert) {
    const methodNames = Object.keys(methods)
    methodNames.forEach(function (method) {
      const methodName = method.toLowerCase().replace(/_\w/g, function (index, match) {
        return index.replace('_', '').toUpperCase()
      })
      assert.isFunction(Response[methodName])
    })
  })

  test('should set response header', async function (assert) {
    const server = http.createServer(function (req, res) {
      Response.header(res, 'content-type', 'application/json')
      res.end()
    })
    await supertest(server).get('/').expect(200).expect('Content-Type', /json/)
  })

  test('should not set response header when already there by using safeHeader method', async function (assert) {
    const server = http.createServer(function (req, res) {
      Response.header(res, 'content-type', 'application/json')
      Response.safeHeader(res, 'content-type', 'text/html')
      res.end()
    })
    await supertest(server).get('/').expect(200).expect('Content-Type', /json/)
  })

  test('should set an array of headers for a given field', async function (assert) {
    const server = http.createServer(function (req, res) {
      Response.header(res, 'Foo', ['bar', 'baz'])
      res.end()
    })
    const res = await supertest(server).get('/').expect(200)
    assert.equal(res.headers.foo, 'bar, baz')
  })

  test('should set response status', async function (assert) {
    const server = http.createServer(function (req, res) {
      Response.status(res, 400)
      res.end()
    })
    await supertest(server).get('/').expect(400)
  })

  test('should send text back to client with proper headers', async function (assert) {
    const server = http.createServer(function (req, res) {
      Response.send(req, res, 'hello world')
    })
    const res = await supertest(server).get('/').expect(200)
    assert.equal(res.text, 'hello world')
    assert.equal(res.headers['content-type'], 'text/plain; charset=utf-8')
  })

  test('should send number back to client with proper headers', async function (assert) {
    const server = http.createServer(function (req, res) {
      Response.send(req, res, 1)
    })
    const res = await supertest(server).get('/').expect(200)
    assert.equal(res.text, '1')
    assert.equal(res.headers['content-type'], 'text/plain; charset=utf-8')
  })

  test('should send json back to client with proper headers', async function (assert) {
    const body = {
      name: 'foo',
      age: 22,
      nested: ['foo', 'bar']
    }
    const server = http.createServer(function (req, res) {
      Response.send(req, res, body)
    })
    const res = await supertest(server).get('/').expect(200)
    assert.deepEqual(res.body, body)
    assert.equal(res.headers['content-type'], 'application/json; charset=utf-8')
  })

  test('should send empty response', async function (assert) {
    const body = ''
    const server = http.createServer(function (req, res) {
      Response.send(req, res, body)
    })
    const res = await supertest(server).get('/').expect(200)
    assert.equal(res.text, '')
    assert.equal(res.headers['content-type'], 'text/plain; charset=utf-8')
  })

  test('should send empty response when request body is null', async function (assert) {
    const body = null
    const server = http.createServer(function (req, res) {
      Response.send(req, res, body)
    })
    const res = await supertest(server).get('/').expect(204)
    assert.equal(res.text, '')
  })

  test('should send boolean as response', async function (assert) {
    const body = true
    const server = http.createServer(function (req, res) {
      Response.send(req, res, body)
    })
    const res = await supertest(server).get('/').expect(200)
    assert.equal(res.text, 'true')
    assert.equal(res.headers['content-type'], 'text/plain; charset=utf-8')
  })

  test('should send buffer as response', async function (assert) {
    const body = Buffer.from('hello world', 'utf8')
    const server = http.createServer(function (req, res) {
      Response.send(req, res, body)
    })
    const res = await supertest(server).get('/').expect(200)
    assert.equal(res.body.toString(), 'hello world')
    assert.equal(res.headers['content-type'], 'application/octet-stream; charset=utf-8')
  })

  test('should send empty body for HEAD request', async function (assert) {
    const body = Buffer.from('hello world', 'utf8')
    const server = http.createServer(function (req, res) {
      Response.send(req, res, body)
    })
    const res = await supertest(server).head('/').expect(200)
    assert.equal(res.headers['content-type'], 'application/octet-stream; charset=utf-8')
    assert.deepEqual(res.body, {})
  })

  test('should not override existing status by calling send method', async function (assert) {
    const body = true
    const server = http.createServer(function (req, res) {
      Response.status(res, 204)
      Response.send(req, res, body)
    })
    await supertest(server).get('/').expect(204)
  })

  test('should send valid json response using json as shorthand method', async function (assert) {
    const body = {
      foo: 'bar',
      baz: ['foo', 'bar']
    }
    const server = http.createServer(function (req, res) {
      Response.json(req, res, body)
    })
    const res = await supertest(server).get('/').expect(200)
    assert.deepEqual(res.body, body)
  })

  test('should send jsonp response back to client with default callback', async function (assert) {
    const body = {
      name: 'foo',
      age: 22,
      nested: ['foo', 'bar']
    }
    const response = "/**/ typeof callback === 'function' && callback(" + JSON.stringify(body) + ');'
    const server = http.createServer(function (req, res) {
      Response.jsonp(req, res, body)
    })
    const res = await supertest(server).get('/').expect(200)
    assert.equal(res.text, response)
    assert.equal(res.headers['content-type'], 'text/javascript; charset=utf-8')
    assert.equal(res.headers['x-content-type-options'], 'nosniff')
  })

  test('should send jsonp response back to client with defined callback', async function (assert) {
    const body = {
      name: 'foo',
      age: 22,
      nested: ['foo', 'bar']
    }
    const response = "/**/ typeof foo === 'function' && foo(" + JSON.stringify(body) + ');'
    const server = http.createServer(function (req, res) {
      Response.jsonp(req, res, body, 'foo')
    })
    const res = await supertest(server).get('/').expect(200)
    assert.equal(res.text, response)
    assert.equal(res.headers['content-type'], 'text/javascript; charset=utf-8')
    assert.equal(res.headers['x-content-type-options'], 'nosniff')
  })

  test('should remove irrelevant headers when response status is 204 or 304', async function (assert) {
    const body = 'hello world'
    const server = http.createServer(function (req, res) {
      Response.status(res, 204)
      Response.send(req, res, body)
    })
    const res = await supertest(server).get('/').expect(204)
    assert.equal(res.headers['content-type'], undefined)
    assert.equal(res.headers['content-length'], undefined)
  })

  test('should send file for download with proper headers', async function (assert) {
    const server = http.createServer(function (req, res) {
      Response.download(req, res, path.join(__dirname, './files/hello.txt'))
    })
    const res = await supertest(server).get('/').expect(200)
    assert.equal(res.text.trim(), 'hello world')
    assert.equal(res.headers['content-type'], 'text/plain; charset=UTF-8')
    assert.notEqual(res.headers['last-modified'], undefined)
  })

  test('should send file to be force downloaded with proper headers', async function (assert) {
    const server = http.createServer(function (req, res) {
      Response.attachment(req, res, path.join(__dirname, './files/hello.txt'))
    })
    const res = await supertest(server).get('/').expect(200)
    assert.equal(res.text.trim(), 'hello world')
    assert.equal(res.headers['content-type'], 'text/plain; charset=UTF-8')
    assert.notEqual(res.headers['last-modified'], undefined)
    assert.equal(res.headers['content-disposition'], 'attachment; filename="hello.txt"')
  })

  test('should throw an error when file is not readable', async function (assert) {
    const server = http.createServer(function (req, res) {
      Response.download(req, res, path.join(__dirname, './files/foo.txt'))
    })
    await supertest(server).get('/').expect(404)
  })

  test('should set location header on response', async function (assert) {
    const server = http.createServer(function (req, res) {
      Response.location(res, 'http://localhost')
      Response.send(req, res, '')
    })
    const res = await supertest(server).get('/').expect(200)
    assert.equal(res.headers['location'], 'http://localhost')
  })

  test('should redirect to a given url', async function (assert) {
    const server = http.createServer(function (req, res) {
      Response.redirect(req, res, 'http://localhost', 301)
    })
    const res = await supertest(server).get('/').expect(301)
    assert.equal(res.headers['location'], 'http://localhost')
    assert.equal(res.headers['content-length'], '0')
  })

  test('should redirect with 302 when status is not defined', async function (assert) {
    const server = http.createServer(function (req, res) {
      Response.redirect(req, res, 'http://localhost')
    })
    const res = await supertest(server).get('/').expect(302)
    assert.equal(res.headers['location'], 'http://localhost')
    assert.equal(res.headers['content-length'], '0')
  })

  test('should add vary header', async function (assert) {
    const server = http.createServer(function (req, res) {
      Response.vary(res, 'Origin')
      Response.send(req, res, '')
    })
    const res = await supertest(server).get('/').expect(200)
    assert.equal(res.headers['vary'], 'Origin')
  })

  test('should not hit the maxListeners when making more than 10 calls', async function (assert) {
    const server = http.createServer(function (req, res) {
      Response.download(req, res, path.join(__dirname, './files/hello.txt'))
    })
    const requests = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13].map(() => supertest(server).get('/').expect(200))
    await Promise.all(requests)
  })

  test('should be able to set content type using the type method', async function (assert) {
    const server = http.createServer(function (req, res) {
      Response.type(res, 'json')
      Response.send(req, res, '')
    })
    const res = await supertest(server).get('/').expect(200)
    assert.equal(res.headers['content-type'], 'application/json; charset=utf-8')
  })

  test('should be able to override the charset using the type method', async function (assert) {
    const server = http.createServer(function (req, res) {
      Response.type(res, 'json', 'myjson')
      Response.send(req, res, '')
    })
    const res = await supertest(server).get('/').expect(200)
    assert.equal(res.headers['content-type'], 'application/json; charset=myjson')
  })

  test('should append to existing headers', async function (assert) {
    const server = http.createServer(function (req, res) {
      Response.header(res, 'Foo', 'bar')
      Response.header(res, 'Foo', 'baz')
      Response.header(res, 'Foo', 'foo')
      res.end()
    })
    const res = await supertest(server).get('/').expect(200)
    assert.equal(res.headers.foo, 'bar, baz, foo')
  })
})


src/Response/index.js
'use strict'

/*
 * node-res
 *
 * (c) Harminder Virk <virk@adonisjs.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
*/
const mime = require('mime-types')
const etag = require('etag')
const contentDisposition = require('content-disposition')
const contentType = require('content-type')
const vary = require('vary')
const send = require('send')
const methods = require('./methods')

const returnContentAndType = function (body) {
  body = body || ''

  /**
   * Return the body and it's type when
   * body is a string.
   */
  if (typeof (body) === 'string') {
    return {
      body,
      type: /^\s*</.test(body) ? 'html' : 'text'
    }
  }

  /**
   * If body is a buffer, return the exact copy
   * and type as bin.
   */
  if (Buffer.isBuffer(body)) {
    return { body, type: 'bin' }
  }

  /**
   * If body is a number or boolean. Convert it to
   * a string and return the type as text.
   */
  if (typeof (body) === 'number' || typeof (body) === 'boolean') {
    return { body: String(body), type: 'text' }
  }

  /**
   * Otherwise check whether body is an object or not. If yes
   * stringify it and otherwise return the exact copy.
   */
  return typeof (body) === 'object' ? { body: JSON.stringify(body), type: 'json' } : { body }
}

/**
 * This is a static utility class to make HTTP response
 * in Node.js. It works like a facade over HTTP res
 * object but without side-effects.
 *
 * @class Response
 * @static
 */
const Response = exports = module.exports = {}

Response.descriptiveMethods = Object.keys(methods).map((method) => {
  const methodName = method.toLowerCase().replace(/_\w/g, function (index, match) {
    return index.replace('_', '').toUpperCase()
  })
  Response[methodName] = function (req, res, body) {
    Response.status(res, methods[method])
    Response.send(req, res, body)
  }
  return methodName
})

/**
 * Returns the value of an existing header on
 * the response object
 *
 * @method getHeader
 *
 * @param  {Object}  res
 * @param  {String}  key
 *
 * @return {Array}
 *
 * @example
 * ```js
 * nodeRes.getHeader(res, 'Content-type')
 * ```
 */
Response.getHeader = function (res, key) {
  return res.getHeader(key)
}

/**
 * Sets header on the response object
 *
 * @method header
 *
 * @param  {Object} res
 * @param  {String} key
 * @param  {String} value
 *
 * @return {void}
 *
 * @example
 * ```js
 * nodeRes.header(res, 'Content-type', 'application/json')
 *
 * // or set an array of headers
 * nodeRes.header(res, 'Link', ['<http://localhost/>', '<http://localhost:3000/>'])
 * ```
 */
Response.header = function (res, key, value) {
  value = Array.isArray(value) ? value : [value]

  /**
   * Fetch existing headers, so that we can append
   * to them
   */
  const previousHeader = Response.getHeader(res, key)
  if (!previousHeader) {
    res.setHeader(key, value.map(String))
    return
  }

  const headers = Array.isArray(previousHeader) ? previousHeader.concat(value) : [previousHeader].concat(value)
  res.setHeader(key, headers.map(String))
}

/**
 * Set status on the HTTP res object
 *
 * @method status
 *
 * @param  {Object} res
 * @param  {Number} code
 *
 * @return {void}
 *
 * @example
 * ```js
 * nodeRes.status(200)
 * ```
 */
Response.status = function (res, code) {
  res.statusCode = code
}

/**
 * Sets the header on response object, only if it
 * does not exists.
 *
 * @method safeHeader
 *
 * @param  {Object}   res
 * @param  {String}   key
 * @param  {String}   value
 *
 * @return {void}
 *
 * @example
 * ```js
 * nodeRes.safeHeader(res, 'Content-type', 'application/json')
 * ```
 */
Response.safeHeader = function (res, key, value) {
  if (!res.getHeader(key)) {
    Response.header(res, key, value)
  }
}

/**
 * @description removing header using it's key
 * @method removeHeader
 * @param  {Object}     res
 * @param  {String}     key
 * @return {void}
 */
Response.removeHeader = function (res, key) {
  res.removeHeader(key)
}

/**
 * Write string or buffer to the response object.
 *
 * @method write
 *
 * @param  {Object} res
 * @param  {String|Buffer} body
 *
 * @return {void}
 *
 * @example
 * ```js
 * nodeRes.write(res, 'Hello world')
 * ```
 */
Response.write = function (res, body) {
  res.write(body)
}

/**
 * Explictly end HTTP response
 *
 * @method end
 *
 * @param  {Object} res
 *
 * @return {void}
 */
Response.end = function (res) {
  res.end()
}

/**
 * Send body as the HTTP response and end it. Also
 * this method will set the appropriate `Content-type`
 * and `Content-length`.
 *
 * If body is set to null, this method will end the response
 * as 204.
 *
 * @method send
 *
 * @param  {Object} req
 * @param  {Object} res
 * @param  {Mixed} body
 *
 * @return {void}
 *
 * @example
 * ```js
 * nodeRes.send(req, res, 'Hello world')
 *
 * // or html
 * nodeRes.send(req, res, '<h2> Hello world </h2>')
 *
 * // or JSON
 * nodeRes.send(req, res, { greeting: 'Hello world' })
 *
 * // or Buffer
 * nodeRes.send(req, res, Buffer.from('Hello world', 'utf-8'))
 * ```
 */
Response.send = function (req, res, body) {
  if (body === null) {
    Response.status(res, 204)
    Response.removeHeader(res, 'Content-Type')
    Response.removeHeader(res, 'Content-Length')
    Response.removeHeader(res, 'Transfer-Encoding')
    Response.end(res)
    return
  }

  let { body: chunk, type } = returnContentAndType(body)
  Response.type(res, type)

  /**
   * setting up content length on response headers
   */
  if (chunk) {
    if (!Buffer.isBuffer(chunk)) {
      chunk = Buffer.from(chunk, 'utf-8')
    }
    const length = chunk.length
    Response.header(res, 'Content-Length', length)
  }

  /**
   * generating and setting etag
   */
  Response.safeHeader(res, 'ETag', etag(chunk))

  /**
   * removing unneccessary headers if response
   * has certain statusCode
   */
  if (res.statusCode === 204 || res.statusCode === 304) {
    Response.removeHeader(res, 'Content-Type')
    Response.removeHeader(res, 'Content-Length')
    Response.removeHeader(res, 'Transfer-Encoding')
  }

  if (req.method !== 'HEAD') {
    Response.write(res, chunk)
  }
  Response.end(res)
}

/**
 * Returns the HTTP response with `Content-type`
 * set to `application/json`.
 *
 * @method json
 *
 * @param  {Object} req
 * @param  {Object} res
 * @param  {Object} body
 *
 * @return {void}
 *
 * @example
 * ```js
 * nodeRes.json(req, res, { name: 'virk' })
 * nodeRes.json(req, res, [ 'virk', 'joe' ])
 * ```
 */
Response.json = function (req, res, body) {
  Response.type(res, 'application/json')
  Response.send(req, res, body)
}

/**
 * Make JSONP response with `Content-type` set to
 * `text/javascript`.
 *
 * @method jsonp
 *
 * @param  {Object}   req
 * @param  {Object}   res
 * @param  {Object}   body
 * @param  {String}   [callbackFn = 'callback']
 *
 * @return {void}
 *
 * @example
 * ```js
 * nodeRes.jsonp(req, res, { name: 'virk' }, 'callback')
 * ```
 */
Response.jsonp = function (req, res, body, callbackFn = 'callback') {
  Response.header(res, 'X-Content-Type-Options', 'nosniff')
  Response.type(res, 'text/javascript')

  body = JSON.stringify(body)

  /**
   * replacing non-allowed javascript characters from body
   */
  body = body
    .replace(/\u2028/g, '\\u2028')
    .replace(/\u2029/g, '\\u2029')

  /**
   * setting up callbackFn on response body , typeof will make
   * sure not to throw error of client if callbackFn is not
   * a function
   */
  body = '/**/ typeof ' + callbackFn + " === 'function' && " + callbackFn + '(' + body + ');'
  Response.send(req, res, body)
}

/**
 * Download file as a stream. Stream will be closed once
 * download is finished.
 *
 * Options are passed directly to [send](https://www.npmjs.com/package/send)
 *
 * @method download
 *
 * @param  {Object} req
 * @param  {Object} res
 * @param  {String} filePath
 * @param  {Object} [options = {}]
 *
 * @return {void}
 *
 * @example
 * ```js
 * nodeRes.download(req, res, '/storage/data.txt')
 * ```
 */
Response.download = function (req, res, filePath, options = {}) {
  send(req, filePath, options).pipe(res)
}

/**
 * Send file as a stream with Content-Disposition of attachment
 * which forces the download of the file.
 *
 * @method attachment
 *
 * @param  {Object}   req
 * @param  {Object}   res
 * @param  {String}   filePath
 * @param  {String}   [name = filePath]
 * @param  {String}   [disposition = 'attachment']
 * @param  {Object}   [options]
 *
 * @return {void}
 *
 * @example
 * ```js
 * nodeRes.attachment(req, res, '/storage/data.txt', 'data.txt')
 * ```
 */
Response.attachment = function (req, res, filePath, name = filePath, disposition = 'attachment', options) {
  Response.header(res, 'Content-Disposition', contentDisposition(name, {type: disposition}))
  send(req, filePath, options).pipe(res)
}

/**
 * Set `Location` header on the HTTP response.
 *
 * @method location
 *
 * @param  {Object} res
 * @param  {String} url
 *
 * @return {void}
 */
Response.location = function (res, url) {
  Response.header(res, 'Location', url)
}

/**
 * Redirect the HTTP request to the given url.
 *
 * @method redirect
 *
 * @param  {Object} req
 * @param  {Object} res
 * @param  {String} url
 * @param  {Number} [status = 302]
 *
 * @return {void}
 *
 * @example
 * ```js
 * nodeRes.redirect(req, res, '/')
 * ```
 */
Response.redirect = function (req, res, url, status = 302) {
  const body = ''
  Response.status(res, status)
  Response.location(res, url)
  Response.header(res, 'Content-Length', Buffer.byteLength(body))
  Response.send(req, res, body)
}

/**
 * Add vary header to the HTTP response.
 *
 * @method vary
 *
 * @param  {Object} res
 * @param  {String} field
 *
 * @return {void}
 */
Response.vary = function (res, field) {
  vary(res, field)
}

/**
 * Set content type header by looking up the actual
 * type and setting charset to utf8
 *
 * @method type
 *
 * @param  {Object} res
 * @param  {String} type
 * @param  {String} [charset = 'utf-8']
 * @return {void}
 *
 * @example
 * ```js
 * nodeRes.type(res, 'html')
 * nodeRes.type(res, 'json')
 * nodeRes.type(res, 'application/json')
 * ```
 */
Response.type = function (res, type, charset = 'utf-8') {
  const ct = type.indexOf('/') === -1 ? mime.lookup(type) || 'text/html' : type
  const parsedType = contentType.parse(ct)
  parsedType.parameters.charset = charset
  Response.safeHeader(res, 'Content-Type', contentType.format(parsedType))
}
